certora/specs/Gates.spec

```diff
@@ -0,0 +1,146 @@
+persistent ghost mapping(address => bool) ghostStatusCanReceiveShares;
+persistent ghost mapping(address => bool) ghostStatusCanSendAssets;
+persistent ghost mapping(address => bool) ghostStatusCanReceiveAssets;
+persistent ghost mapping(address => bool) ghostBalanceChangeAllowed;
+
+hook Sstore ERC20.balanceOf[KEY address user] uint256 newBalance (uint256 oldBalance) {
+    if (!canReceiveAssets(user)) {
+        ghostBalanceChangeAllowed[user] = oldBalance >= newBalance && ghostBalanceChangeAllowed[user];
+    }
+    if (!canSendAssets(user)) {
+        ghostBalanceChangeAllowed[user] = oldBalance <= newBalance && ghostBalanceChangeAllowed[user];
+    }
+}
+
+rule cantReceiveShares(env e, method f, calldataarg args, address user) {
+    require(currentContract.sharesGate != 0, "require gating to be enabled for shares");
+
+    // Trick to require that all the following addresses are different.
+    require(MorphoMarketV1Adapter == 0x10, "ack");
+    require(MorphoVaultV1Adapter == 0x11, "ack");
+    require(currentContract == 0x12, "ack");
+
+    require (!canReceiveShares(user), "require that the user under scrutiny can't receive shares");
+
+    uint256 sharesBefore = balanceOf(user);
+
+    f(e, args);
+
+    assert balanceOf(user) <= sharesBefore;
+}
+
+rule cantSendShares(env e, method f, calldataarg args, address user, uint256 shares) {
+    require(currentContract.sharesGate != 0, "require gating to be enabled for shares");
+
+    // Trick to require that all the following addresses are different.
+    require(MorphoMarketV1Adapter == 0x10, "ack");
+    require(MorphoVaultV1Adapter == 0x11, "ack");
+    require(currentContract == 0x12, "ack");
+
+    require (!canSendShares(user), "require that the user under scrutiny can't send shares");
+
+    uint256 sharesBefore = balanceOf(user);
+
+    f(e, args);
+
+    assert balanceOf(user) >= sharesBefore;
+}
+
+rule cantSendAssetsAndCantReceiveAssets(env e, method f, calldataarg args, address user)  filtered {
+    f -> f.selector != sig:MorphoMarketV1Adapter.skim(address).selector &&
+         f.selector != sig:MorphoVaultV1Adapter.skim(address).selector
+}{
+    require(currentContract.sendAssetsGate != 0, "setup gating");
+    require (!canSendAssets(user), "setup gating");
+
+    // Trick to require that all the following addresses are different.
+    require(MorphoMarketV1Adapter == 0x10, "ack");
+    require(MorphoVaultV1Adapter == 0x11, "ack");
+    require(currentContract == 0x12, "ack");
+    require(asset() == 0x13, "ack");
+
+    require (user != MorphoMarketV1Adapter && user != MorphoVaultV1Adapter && user != currentContract, "do not check if the vault or the adapters themselves are properly gated to not send assets");
+    require (forall address adapter . isAdapter(adapter) => (adapter == MorphoMarketV1Adapter || adapter == MorphoVaultV1Adapter), "require that the liquidity adapter is unset or a known implementation");
```

Comment:
The reason is outdated, we could say something like "Only check the property for adapters that have been developed until now"